### PR TITLE
TRUNK-5338 Add changeset to create or modify conditions table

### PR DESCRIPTION
**Pull request by paradisekelechi**
## Description of what I changed

Added liquibase changeset to add a new conditions table or modify an existing one. 

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-5338

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by dkayiwa** [comment]
@paradisekelechi did you actually test this out by running an instance of openmrs, go through the database setup wizard, complete it, check the database to confirm all is changed as expected?

**Comment by paradisekelechi** [comment]
@dkayiwa 
I actually tested it. I ran it an the changes were effected accordingly on the already existing `conditions` table. After that, I dropped the table and ran and checked the other changeset. 

However, I do not know about the database setup wizard.  Could you share resources with respect to that?

**Comment by dkayiwa** [comment]
In the liquibasechangelog table, delete the row with this changeset id. Then restart openmrs and tell me what you get in the browser.

**Comment by paradisekelechi** [comment]
About that, I did that a lot of times until I no longer have errors on the console. 
I'd run that again and update you accordingly.

**Comment by dkayiwa** [comment]
How are you dealing with the data migration? Didn't some columns store data in a different format than  that of the old table?

**Comment by paradisekelechi** [comment]
The only tables that changed are `previous_condition_id`, `status` and `concept_id` and they only changed in name and not datatype.

**Comment by paradisekelechi** [comment]
What I am only concerned about is adding foreign key constraints after converting an old conditions table to the new one. . .

**Comment by dkayiwa** [comment]
1) Can you revert to the old table.
2) Create around 5 conditions from the user interface when running on openmrs-core versions 2.1.3 or lower.
3) Now switch to openmrs-core 2.2.0-SNAPSHOT to have your changesets run
4) Check to see if the data is migrated and expected and also displayed in the UI using the latest snapshot versions of the ermapi and coreapps modules.

File: api/src/main/resources/liquibase-update-to-2.2.xml [hunk 1]
```diff
@@ -117,14 +117,29 @@
...
             referencedTableName="users" referencedColumnNames="user_id" />
     </changeSet>
+	<changeSet id="20180501-TRUNK-5338" author="Kelechi+iheanyi">
+		<preConditions onFail="MARK_RAN">
+			<tableExists tableName="conditions"/>
+		</preConditions>
+		<comment>
+			Modify conditions table to match condition object in the core platform if the conditions table already exists in the database
+		</comment>
+		<renameColumn columnDataType="int" newColumnName="previous_version" oldColumnName="previous_condition_id" tableName="conditions"/>
```
Comment (wluyima) [inline api/src/main/resources/liquibase-update-to-2.2.xml]: I would prefer to have these as separate changesets with their respective comments being explicit about what they do e.g **Rename tableA.x column to y**

File: api/src/main/resources/liquibase-update-to-2.2.xml [hunk 2]
```diff
@@ -117,14 +117,29 @@
...
+		<renameColumn columnDataType="varchar(255)" newColumnName="clinical_status" oldColumnName="status" tableName="conditions"/>
+		<renameColumn columnDataType="int" newColumnName="condition_coded" oldColumnName="concept_id" tableName="conditions"/>
+
+		<addColumn tableName="conditions">
```
Comment (wluyima) [inline api/src/main/resources/liquibase-update-to-2.2.xml]: This should also be a separate changeset

File: api/src/main/resources/liquibase-update-to-2.2.xml [hunk 3]
```diff
@@ -117,14 +117,29 @@
...
+			<column name="condition_coded_name" type="int"/>
+		</addColumn>
+	</changeSet>
 	<changeSet id="20180214-1000" author="Kelechi+iheanyi">
```
Comment (wluyima) [inline api/src/main/resources/liquibase-update-to-2.2.xml]: It's good to change the id when you change the changeset contents so that this doesn't break other developer's working copies or our CI servers

File: api/src/main/resources/liquibase-update-to-2.2.xml [hunk 4]
```diff
@@ -117,14 +117,29 @@
...
 		<preConditions onFail="MARK_RAN">
-			<not><tableExists tableName="condition"/></not>
+			<not><tableExists tableName="conditions"/></not>
 		</preConditions>
 		<comment>
-			Add condition table
+			Add conditions table if the table does not exist in the database
```
[the remark this answers is not in the document]
Comment (paradisekelechi) [reply api/src/main/resources/liquibase-update-to-2.2.xml]: Okay. Noted

File: api/src/main/resources/liquibase-update-to-2.2.xml (hunk 3 above)
Comment (paradisekelechi) [reply api/src/main/resources/liquibase-update-to-2.2.xml]: Yes. Done

File: api/src/main/resources/liquibase-update-to-2.2.xml (hunk 2 above)
Comment (paradisekelechi) [reply api/src/main/resources/liquibase-update-to-2.2.xml]: Okay. Noted

File: api/src/main/resources/liquibase-update-to-2.2.xml (hunk 1 above)
Comment (paradisekelechi) [reply api/src/main/resources/liquibase-update-to-2.2.xml]: Okay. Done

File: api/src/main/resources/liquibase-update-to-2.2.xml (hunk 3 above)
Comment (wluyima) [reply api/src/main/resources/liquibase-update-to-2.2.xml]: I don't see the change

**Comment by dkayiwa** [comment]
@paradisekelechi have you tried to run a fresh setup of openmrs with these changes?

**Comment by dkayiwa** [comment]
@paradisekelechi if you do, you should get this error: https://pastebin.com/QxxXu5h6

**Comment by paradisekelechi** [comment]
@dkayiwa
I have run a fresh setup of the core platform following the steps you highlighted. I did not get that error. The changesets ran successfully without errors

**Comment by dkayiwa** [comment]
Do you have the reference application modules loaded?

**Comment by paradisekelechi** [comment]
Yes I do have the reference application modules loaded

**Comment by dkayiwa** [comment]
That does not look like a fresh setup. For the emrapi module would attempt to create a conditions table which should fail because the core platform already created it.

**Comment by paradisekelechi** [comment]
@dkayiwa 
I have done the above. This is the output I get in the console for these changesets: https://pastebin.com/Fr0w4nt9 .
On checking the db created, the `conditions` table is there with the changes effected accordingly.

**Comment by paradisekelechi** [comment]
1. I dropped the previous databases I had.
2. I deleted the openmrs-runtime.properties file
3. I started the server and followed through on the steps which created a new db and updated it by running the changesets

**Comment by dkayiwa** [comment]
As you did that, which version of openmrs-core were you running?

**Comment by paradisekelechi** [comment]
2.2.0-SNAPSHOT

**Comment by dkayiwa** [comment]
And when you run the setup, did you have all the reference application modules loaded from the very start?

**Comment by paradisekelechi** [comment]
No I did not have them

**Comment by dkayiwa** [comment]
@paradisekelechi can you do that?

**Comment by dkayiwa** [comment]
Now do a fresh setup with those modules installed

**Comment by samuelmale** [comment]
Are you still working on this @paradisekelechi ?

**Comment by samuelmale** [comment]
Are you still working on this @paradisekelechi ?

**Comment by paradisekelechi** [comment]
I have worked on this. I am waiting for verifications and updates from @dkayiwa

File: api/src/main/resources/liquibase-update-to-2.2.xml [hunk 5]
```diff
@@ -117,14 +117,23 @@
...
     </changeSet>
-	<changeSet id="20180214-1000" author="Kelechi+iheanyi">
+	<changeSet author="Kelechi+iheanyi" id="20180508-1000">
 		<preConditions onFail="MARK_RAN">
-			<not><tableExists tableName="condition"/></not>
+			<tableExists tableName="conditions"/>
 		</preConditions>
 		<comment>
-			Add condition table
+			Rename conditions table to emrapi_conditions
```
Comment (wluyima) [inline api/src/main/resources/liquibase-update-to-2.2.xml]: This comments seems to be saying the opposite

Comment (wluyima) [inline api/src/main/resources/liquibase-update-to-2.2.xml]: Why are you renaming the table, I thought you should be updating it to match the new design

Comment (paradisekelechi) [reply api/src/main/resources/liquibase-update-to-2.2.xml]: @wluyima 
We are renaming the table because of some edge cases we saw in the implementation. 
Assume we are running a fresh instance of the core platform and the emrapi module.
If we update the `conditions` table here, the emrapi module tries to create a new `conditions` table which throws errors.

However, the following implementation handles that edge case successfully. On the core platform, the conditions table is renamed to emrapi_conditions. Then a conditions table is created and data migrated to it. On the emrapi module, the emrapi_conditions table is used.

**Comment by paradisekelechi** [comment]
I have thoroughly tested this @dkayiwa

File: api/src/main/resources/liquibase-update-to-2.2.xml (hunk 5 above)
Comment (wluyima) [reply api/src/main/resources/liquibase-update-to-2.2.xml]: You should not be renaming the table, in any case it's no longer going to live in the emrapi module so there if no need to add that prefix

**Comment by samuelmale** [comment]
Hi @paradisekelechi 
Are still working on this? Is so, did you address all @wluyima's  requested changes? 
Please look into this and we close this :)

File: api/src/main/resources/liquibase-update-to-2.2.xml (hunk 5 above)
Comment (mozzy11) [reply api/src/main/resources/liquibase-update-to-2.2.xml]: @wluyima  , wat do u mean if the table exists , just do the migration ..? i think that will be just do the modification.  because migration is posiible only after first renaming it.
